File: app/src/main/java/com/education4all/MathCoachAlg/StatisticMaker.java
package com.education4all.MathCoachAlg;

import android.content.Context;
import android.content.SharedPreferences;

import com.education4all.MathCoachAlg.Tours.Tour;

import java.util.ArrayList;


public class StatisticMaker {

    public static final String STATISTICS = "Statistics";
    public static final String TOURS = "tours";


    public static void saveTour(Tour tour, Context context) {
        SharedPreferences.Editor editor = context.getSharedPreferences(STATISTICS, Context.MODE_PRIVATE).edit();
        SharedPreferences prefs = context.getSharedPreferences(STATISTICS, Context.MODE_PRIVATE);

        String tourCountStr = prefs.getString(TOURS, "0");
        int tourCount = Integer.parseInt(tourCountStr);
        saveTour_by_number(tour, context, tourCount);
        editor.putString(TOURS, Integer.toString(tourCount + 1));
        editor.apply();
    }

    public static void saveTourOLD(Tour tour, Context context) {
        SharedPreferences.Editor l_editor = context.getSharedPreferences(STATISTICS, Context.MODE_PRIVATE).edit();
        SharedPreferences prefs = context.getSharedPreferences(STATISTICS, Context.MODE_PRIVATE);
        String tourCountStr = prefs.getString(TOURS, "0");
        int tourCount = Integer.parseInt(tourCountStr);
        l_editor.putString(TOURS + "_" + tourCountStr, Integer.toString(tour.getTotalTasks()));
        ArrayList<String> serializedTour = tour.serializeOLD();
        for (int taskNumber = 0; taskNumber < tour.getTotalTasks() + 2; ++taskNumber) {
            l_editor.putString(TOURS + "_" + tourCountStr + "_" + taskNumber, serializedTour.get(taskNumber));
        }
        l_editor.putString(TOURS, Integer.toString(tourCount + 1));
        l_editor.apply();
    }

    public static ArrayList<Tour> loadToursOLD(Context context) {
        ArrayList<Tour> resultTours = new ArrayList<>();
        SharedPreferences prefs = context.getSharedPreferences(STATISTICS, Context.MODE_PRIVATE);
        int toursCount = Integer.parseInt(prefs.getString(TOURS, "0"));

        for (int tourNumber = 0; tourNumber < toursCount; ++tourNumber)
            resultTours.add(loadTourOld(context, tourNumber));

        return resultTours;
    }

    public static Tour loadTour(Context context, int tourNumber) {
        SharedPreferences prefs = context.getSharedPreferences(STATISTICS, Context.MODE_PRIVATE);
        String currentTour = prefs.getString(TOURS + "_" + tourNumber, "");
        return Tour.deSerialize(currentTour);
    }

    public static Tour loadTourOld(Context context, int tourNumber) {
        SharedPreferences prefs = context.getSharedPreferences(STATISTICS, Context.MODE_PRIVATE);
        int taskCount = Integer.parseInt(prefs.getString(TOURS + "_" + tourNumber, "0"));
        ArrayList<String> currentTour = new ArrayList<>();
        for (int taskNumber = 0; taskNumber < taskCount + 2; ++taskNumber) {
            currentTour.add(prefs.getString(TOURS + "_" + tourNumber + "_" + taskNumber, "0"));
        }
        if (currentTour.size() > 0) {
            return new Tour(currentTour);
        } else {
            return null;
        }
    }

    public static int getTourCount(Context context) {
        SharedPreferences prefs = context.getSharedPreferences(STATISTICS, Context.MODE_PRIVATE);
        return Integer.parseInt(prefs.getString(TOURS, "0"));
    }

    public static String getTourInfoOLD(Context context, int tourNumber) {
        SharedPreferences prefs = context.getSharedPreferences(STATISTICS, Context.MODE_PRIVATE);
        return prefs.getString(TOURS + "_" + tourNumber + "_" + "0", "");
    }

    public static void removeStatistics(Context context) {
        SharedPreferences.Editor editor = context.getSharedPreferences(STATISTICS, Context.MODE_PRIVATE).edit();
        editor.clear();
        editor.apply();
    }

    public static void removeTour(Context context, int tourNumber) {
        SharedPreferences.Editor editor = context.getSharedPreferences(STATISTICS, Context.MODE_PRIVATE).edit();
        SharedPreferences prefs = context.getSharedPreferences(STATISTICS, Context.MODE_PRIVATE);
        editor.remove(TOURS + "_" + tourNumber);
        int toursCount = Integer.parseInt(prefs.getString(TOURS, "0"));
        editor.putString(TOURS, Integer.toString(toursCount - 1));
        editor.apply();
        for (int i = tourNumber + 1; i < toursCount; ++i) {
            Tour tour = loadTour(context, i);
            saveTour_by_number(tour, context, i - 1);
        }
    }

    public static void removeTourOld(Context context, int tourNumber) {
        SharedPreferences.Editor editor = context.getSharedPreferences(STATISTICS, Context.MODE_PRIVATE).edit();
        SharedPreferences prefs = context.getSharedPreferences(STATISTICS, Context.MODE_PRIVATE);
        int taskCount = Integer.parseInt(prefs.getString(TOURS + "_" + tourNumber, "0"));
        for (int taskNumber = 0; taskNumber < taskCount + 2; ++taskNumber) {
            editor.remove(TOURS + "_" + tourNumber + "_" + taskNumber);
        }
        editor.remove(TOURS + "_" + tourNumber);
        int toursCount = Integer.parseInt(prefs.getString(TOURS, "0"));
        editor.putString(TOURS, Integer.toString(toursCount - 1));
        editor.apply();
        for (int i = tourNumber + 1; i < toursCount; ++i) {
            Tour tour = loadTour(context, i);
            saveTour_by_number(tour, context, i - 1);
        }
    }

    public static void saveTour_by_number(Tour tour, Context context, int tourNumber) {
        SharedPreferences.Editor editor = context.getSharedPreferences(STATISTICS, Context.MODE_PRIVATE).edit();
        String tourNumberString = Integer.toString(tourNumber);
        String serializedTour = tour.serialize();
        editor.putString(TOURS + "_" + tourNumberString, serializedTour);
        editor.apply();
    }
}

File: app/src/main/java/com/education4all/MathCoachAlg/StatisticMaker.kt
package com.education4all.mathCoachAlg

import android.content.*
import com.education4all.firebase.FireBaseUtils
import com.education4all.mathCoachAlg.tours.Tour

object StatisticMaker {
    const val STATISTICS = "Statistics"
    const val TOURS = "tours"
    fun saveTour(tour: Tour?, context: Context) {
        val editor = context.getSharedPreferences(STATISTICS, Context.MODE_PRIVATE).edit()
        val prefs = context.getSharedPreferences(STATISTICS, Context.MODE_PRIVATE)
        val tourCountStr = prefs.getString(TOURS, "0")
        val tourCount = tourCountStr!!.toInt()
        saveTour_by_number(tour, context, tourCount)
        editor.putString(TOURS, Integer.toString(tourCount + 1))
        editor.apply()
    }

    fun saveTourOLD(tour: Tour?, context: Context) {
        val l_editor = context.getSharedPreferences(STATISTICS, Context.MODE_PRIVATE).edit()
        val prefs = context.getSharedPreferences(STATISTICS, Context.MODE_PRIVATE)
        val tourCountStr = prefs.getString(TOURS, "0")
        val tourCount = tourCountStr!!.toInt()
        l_editor.putString(TOURS + "_" + tourCountStr, Integer.toString(tour.getTotalTasks()))
        val serializedTour = tour!!.serializeOLD()
        for (taskNumber in 0 until tour.totalTasks + 2) {
            l_editor.putString(
                TOURS + "_" + tourCountStr + "_" + taskNumber,
                serializedTour!![taskNumber]
            )
        }
        l_editor.putString(TOURS, Integer.toString(tourCount + 1))
        l_editor.apply()
    }

    fun loadToursOLD(context: Context): ArrayList<Tour?> {
        val resultTours = ArrayList<Tour?>()
        val prefs = context.getSharedPreferences(STATISTICS, Context.MODE_PRIVATE)
        val toursCount = prefs.getString(TOURS, "0")!!.toInt()
        for (tourNumber in 0 until toursCount) resultTours.add(loadTourOld(context, tourNumber))
        return resultTours
    }

    fun loadTour(context: Context, tourNumber: Int): Tour? {
        val prefs = context.getSharedPreferences(STATISTICS, Context.MODE_PRIVATE)
        val currentTour = prefs.getString(TOURS + "_" + tourNumber, "")
        return Tour.Companion.deSerialize(currentTour)
    }

    fun loadTourOld(context: Context, tourNumber: Int): Tour? {
        val prefs = context.getSharedPreferences(STATISTICS, Context.MODE_PRIVATE)
        val taskCount = prefs.getString(TOURS + "_" + tourNumber, "0")!!.toInt()
        val currentTour = ArrayList<String?>()
        for (taskNumber in 0 until taskCount + 2) {
            currentTour.add(prefs.getString(TOURS + "_" + tourNumber + "_" + taskNumber, "0"))
        }
        return if (currentTour.size > 0) {
            Tour(currentTour)
        } else {
            null
        }
    }

    fun getTourCount(context: Context): Int {
        val prefs = context.getSharedPreferences(STATISTICS, Context.MODE_PRIVATE)
        return prefs.getString(TOURS, "0")!!.toInt()
    }

    fun getTourInfoOLD(context: Context, tourNumber: Int): String? {
        val prefs = context.getSharedPreferences(STATISTICS, Context.MODE_PRIVATE)
        return prefs.getString(TOURS + "_" + tourNumber + "_" + "0", "")
    }

    fun removeStatistics(context: Context) {
        val editor = context.getSharedPreferences(STATISTICS, Context.MODE_PRIVATE).edit()
        editor.clear()
        editor.apply()
        FireBaseUtils().deleteTours()
    }

    fun removeTour(context: Context, tourNumber: Int) {
        val editor = context.getSharedPreferences(STATISTICS, Context.MODE_PRIVATE).edit()
        val prefs = context.getSharedPreferences(STATISTICS, Context.MODE_PRIVATE)
        editor.remove(TOURS + "_" + tourNumber)
        val toursCount = prefs.getString(TOURS, "0")!!.toInt()
        editor.putString(TOURS, Integer.toString(toursCount - 1))
        editor.apply()
        for (i in tourNumber + 1 until toursCount) {
            val tour = loadTour(context, i)
            saveTour_by_number(tour, context, i - 1)
        }
    }

    fun removeTourOld(context: Context, tourNumber: Int) {
        val editor = context.getSharedPreferences(STATISTICS, Context.MODE_PRIVATE).edit()
        val prefs = context.getSharedPreferences(STATISTICS, Context.MODE_PRIVATE)
        val taskCount = prefs.getString(TOURS + "_" + tourNumber, "0")!!.toInt()
        for (taskNumber in 0 until taskCount + 2) {
            editor.remove(TOURS + "_" + tourNumber + "_" + taskNumber)
        }
        editor.remove(TOURS + "_" + tourNumber)
        val toursCount = prefs.getString(TOURS, "0")!!.toInt()
        editor.putString(TOURS, Integer.toString(toursCount - 1))
        editor.apply()
        for (i in tourNumber + 1 until toursCount) {
            val tour = loadTour(context, i)
            saveTour_by_number(tour, context, i - 1)
        }
    }

    fun saveTour_by_number(tour: Tour?, context: Context, tourNumber: Int) {
        val editor = context.getSharedPreferences(STATISTICS, Context.MODE_PRIVATE).edit()
        val tourNumberString = Integer.toString(tourNumber)
        val serializedTour = tour!!.serialize()
        editor.putString(TOURS + "_" + tourNumberString, serializedTour)
        editor.apply()
        FireBaseUtils().uploadTour(tour)
    }
}